### DISCUSSION TITLE
Unary operator overlaps with identifier -- dynamic lexical precedence?

### DISCUSSION THREAD

**Discussion by joaodefelipe** [tree-sitter/tree-sitter#2761] [Q&A]
# Reproducing the issue

I'm trying to write the first building blocks of a deceptively simple grammar. Omitting the part concerning how expressions are separated (as it doesn't matter here), this is my attempt at parsing a single expression:
```javascript
module.exports = grammar({
  name: 'test',
  extras: $ => [],
  rules: {
    expression: $ => seq(
      optional($.prefix),
      $.identifier,
    ),
    prefix: $ => /=/,
    identifier: $ => /\S+/,
  }
});
```
The idea is to match the behavior of the regex `(=?)(\S+)`. A single `=` should parse as
```
(expression [0, 0] - [0, 1]
  (identifier [0, 0] - [0, 1])
```
... but this happens:
```
(ERROR [0, 0] - [0, 1]
  (prefix [0, 0] - [0, 1]))
```
A `==` should be:
```
(expression [0, 0] - [0, 2]
  (prefix [0, 0] - [0, 1])
  (identifier [0, 1] - [0, 2]))
``` 
... but this happens:
```
(expression [0, 0] - [0, 2]
  (identifier [0, 0] - [0, 2]))
```

# Attempts

## Lexical precedence

Setting a greater lexical precedence for the prefix (`prefix: $ => token(prec(1, /\S+/))`) parses `==` correctly, but not `=`. Setting a greater lexical precedence for the identifier (`identifier: $ => token(prec(1, /=/))`) parses `=` correctly, but not `==`.

## Dynamic precedence

I've read the docs over and over, but didn't "get" it. It behaves the same as the first example.
```javascript
module.exports = grammar({
  name: 'test',
  extras: $ => [],
  conflicts: $ => [[$._expression, $._prefixed_expression]],
  rules: {
    expression: $ => choice(prec.dynamic(1, $._prefixed_expression), $._expression),
    _prefixed_expression: $ => seq($.prefix, $.identifier),
    _expression: $ => $.identifier,
    prefix: $ => /=/,
    identifier: $ => /\S+/,
  }
});
```

# Thoughts

The prefix *must* have greater lexical precedence when it is followed by an identifier, but the identifier *must* have lexical precedence when it is not preceded by a prefix. My understanding is that the error comes from this being a chicken-and-egg problem -- setting the lexer precedence for prefixes highest requires backtracking with the expression `=` after the first character is consumed and chosen as a prefix, which is not possible.

I considered leveraging the fact expressions are separated by `/\s+/` and spamming `token.immediate` -- however, to evolve the grammar to its final form, I'd need to change the prefix to something like `choice(/=/, /:\w+/)`, and the identifier to `choice(/\S+/, /'[^']*'/)`. In it, because `:foo` is a valid prefix, one character of lookahead is not sufficient. With such a prefix, a quoted string becomes the only choice for a valid expression. This means `:foo` is just an identifier, but `:foo'bar'` is the prefix `:foo'` followed by the identifier `'bar`.

What's most astonishing to me is that the behavior my initial grammar tries to replicate is just the regex `(=?)(\S+)`, and the more complex behavior described on the previous paragraph is just `(=|:\w+)?(\S*|'[^']*')`. This is starting to convince me that such a behavior is not achiveable without backtracking. Even then, the "monolithic" `'foo'` token would have to be broken to match what's inside the quotes, like on #1297.

The cleanest solution would be to instruct the parser to jump straight to step 5 on the [conflict resolution checklist](https://tree-sitter.github.io/tree-sitter/creating-parsers#conflicting-tokens) -- as #1275 showed, this can be done if steps 1 through 4 fail to solve the conflict. However, I can't just turn the entire expression into a token, because a token is atomic -- besides, the prefix is not guaranteed to be the first element of an expression.

What *is* guaranteed is that the identifier *must* be the last element of an expression -- maybe a right-to-left parser is needed for this part? Indeed, using a suffix instead of a prefix works fine (the expression is guaranteed to start with an identifier), but sadly the grammar in question is already frozen -- I can't change it. 

# TL;DR

- My grammar is composed of expressions.
- Expressions contain an identifier, optionally preceded by an unary operator (prefix).
- The operator (prefix) has a variable number and type of characters, that may overlap with the identifier that succeeds it (also with variable number of characters).
- If it's ambiguous where the operator (prefix) ends and the identifier begins, the entire expression should be treated as an identifier.

Is an external scanner implementing these the only way out?

**Comment by wldmr** [comment]
Not sure how interested you still are after almost two years, but while researching a similar problem I stumbled across this. I can get your test cases to work with this grammar:

```js
module.exports = grammar({
  name: 'test',
  extras: _ => [],
  rules: {
    expression: $ => choice(
      alias($.prefix, $.identifier),
      seq(optional($.prefix), $.identifier)
    ),
    prefix: _ => token(prec(1, /=|:\w+/)),
    identifier: _ => /\S+|'[^']*'/,
  }
});
```
So basically, set higher lexical precedence for the prefix, so it always gets selected, and then just bluntly rename the prefix to identifier if it's the only thing that matches.

Not sure if it's in the spirit of the game, but it makes all these tests pass:

<details>

<summary>Test cases</summary>

```
===
 =
===
=
---
(expression
  (identifier))

===
 ==
===
==
---
(expression
  (prefix)
  (identifier))

===
 =word
===
=word
---
(expression
  (prefix)
  (identifier))

===
 hi
===
hi
---
(expression
  (identifier))

===
 'hi di ho'
===
'hi di ho'
---
(expression
  (identifier))

===
 :foo
===
:foo
---
(expression
  (identifier))

===
 :foo'bar'
===
:foo'bar'
---
(expression
  (prefix)
  (identifier))
```
</details>

Does that work for you?

**Comment by RedCMD** [comment]
I can confirm this is still an issue
even a bug?

and from my experience with TreeSitter
I can say this is impossible without an [External Scanner](https://tree-sitter.github.io/tree-sitter/creating-parsers/4-external-scanners.html#external-scanners) 
or tricky hacks like aliasing the rules in precise ways as shown by @wldmr 

as TreeSitter just blindly picks the longest match before asking the user what they wanted
